crates/rattler_repodata_gateway/src/gateway/query.rs

```diff
@@ -117,46 +141,74 @@ impl GatewayQuery {
             });
         }
 
-        // Package names that we have or will issue requests for.
-        let mut seen = HashSet::new();
-        let mut pending_package_specs = HashMap::new();
-        for spec in self.specs {
-            if let Some(name) = &spec.name {
-                seen.insert(name.clone());
-                pending_package_specs
-                    .entry(name.clone())
-                    .or_insert_with(Vec::new)
-                    .push(spec);
-            }
-        }
+        let len = subdirs.len();
+        let mut result = vec![RepoData::default(); len];
 
-        // A list of futures to fetch the records for the pending package names. The main task
-        // awaits these futures.
+        // A list of futures to fetch the records for the pending package names. The
+        // main task awaits these futures.
         let mut pending_records = FuturesUnordered::new();
 
-        // The resulting list of repodata records.
-        let mut result = vec![RepoData::default(); subdirs.len()];
+        // Push the direct url queries to the pending_records.
+        for spec in direct_url_specs.clone() {
+            let gateway = self.gateway.clone();
+            pending_records.push(
+                async move {
+                    let url = spec
+                        .clone()
+                        .url
+                        .ok_or(GatewayError::MatchSpecNoName(spec.clone()))?;
+                    let query = DirectUrlQuery::new(
+                        url.clone(),
+                        gateway.package_cache.clone(),
+                        gateway.client.clone(),
+                    );
+
+                    let record = query
+                        .execute()
+                        .await
+                        .map_err(|e| GatewayError::DirectUrlQueryError(url.to_string(), e))?;
+
+                    // Check if record actually has the same name
+                    if let Some(record) = record.first() {
+                        let spec_name = spec.clone().name.ok_or(MatchSpecNoName(spec.clone()))?;
+                        if record.package_record.name != spec_name {
+                            // Using as_source to get the closest to the retrieved input.
+                            return Err(GatewayError::NotMatchingNameUrl(
+                                record.package_record.name.as_source().to_string(),
+                                spec_name.as_source().to_string(),
+                            ));
+                        }
+                    }
+                    // Push the direct url record into the last subdir
+                    Ok((len - 1, vec![spec], record))
```

Comment:
This should most likely be the first instead of the last.

Reply:
Made it the first